rasterio/env.py

```diff
@@ -0,0 +1,256 @@
+
+        Notes
+        -----
+        In write mode, you must specify at least ``width``, ``height``,
+        ``count`` and ``dtype``.
+
+        A coordinate reference system for raster datasets in write mode
+        can be defined by the ``crs`` argument. It takes Proj4 style
+        mappings like
+
+        .. code::
+
+          {'proj': 'longlat', 'ellps': 'WGS84', 'datum': 'WGS84',
+           'no_defs': True}
+
+        An affine transformation that maps ``col,row`` pixel coordinates
+        to ``x,y`` coordinates in the coordinate reference system can be
+        specified using the ``transform`` argument. The value should be
+        an instance of ``affine.Affine``
+
+        .. code:: python
+
+            >>> from affine import Affine
+            >>> Affine(0.5, 0.0, -180.0, 0.0, -0.5, 90.0)
+
+        These coefficients are shown in the figure below.
+
+        .. code::
+
+          | x |   | a  b  c | | c |
+          | y | = | d  e  f | | r |
+          | 1 |   | 0  0  1 | | 1 |
+
+          a: rate of change of X with respect to increasing column,
+             i.e. pixel width
+          b: rotation, 0 if the raster is oriented "north up"
+          c: X coordinate of the top left corner of the top left pixel
+          d: rotation, 0 if the raster is oriented "north up"
+          e: rate of change of Y with respect to increasing row,
+             usually a negative number (i.e. -1 * pixel height) if
+             north-up.
+          f: Y coordinate of the top left corner of the top left pixel
+
+        A 6-element sequence of the affine transformation matrix
+        coefficients in ``c, a, b, f, d, e`` order, (i.e. GDAL
+        geotransform order) will be accepted until 1.0 (deprecated).
+
+        A virtual filesystem can be specified. The ``vfs`` parameter may
+        be an Apache Commons VFS style string beginning with "zip://" or
+        "tar://"". In this case, the ``path`` must be an absolute path
+        within that container.
+
+        """
+        if not isinstance(path, string_types):
+            raise TypeError("invalid path: {0!r}".format(path))
+        if mode and not isinstance(mode, string_types):
+            raise TypeError("invalid mode: {0!r}".format(mode))
+        if driver and not isinstance(driver, string_types):
+            raise TypeError("invalid driver: {0!r}".format(driver))
+        if dtype and not check_dtype(dtype):
+            raise TypeError("invalid dtype: {0!r}".format(dtype))
+        if transform:
+            transform = guard_transform(transform)
+        elif 'affine' in kwargs:
+            affine = kwargs.pop('affine')
+            transform = guard_transform(affine)
+
+        # If the current environment, either passed env or self, has
+        # been entered, start it now. This will almost always be the
+        # case when calling `rasterio.open()` with no surrounding
+        # environment.
+        _env = env or self
```

Comment:
@sgillies I think we want `rasterio.open()` to be responsible for handling the optional environment parameter rather than `Env.open()`.  At this point we already have an instantiated `Env()` and the user wants to open a datasource directly with that environment.  The idea that they should be allowed to pass in a _different_ environment seems like something we don't want to support.

```python
import rasterio

with rasterio.Env() as env1, rasterio.Env() as env2:
    with env1.open('datasource.tif', env=env2) as src:
        # Using env1 to open a datasource, but don't actually use that env.  Use this other one instead.
```